mass-deployer/pkg/mass-deployer/load.go

```diff
@@ -0,0 +1,173 @@
+	return os.WriteFile(output, outputBytes, 0644)
+}
+
+func loadNodeGroupsInfo(ctx context.Context, tfPluginClient deployer.TFPluginClient, nodegroups []string, maxRetries uint64, output string) ([]byte, error) {
+	trial := 1
+	nodeGroupsInfo := map[string][]vmOutput{}
+	failedGroups := map[string]string{}
+
+	// load contracts with node group name
+	for _, nodeGroup := range nodegroups {
+		if err := retry.Do(ctx, retry.WithMaxRetries(maxRetries, retry.NewConstant(1*time.Second)), func(ctx context.Context) error {
+			if trial != 1 {
+				log.Debug().Str("Node group", nodeGroup).Int("Deployment trial", trial).Msg("Retrying to load")
+			}
+
+			contracts, err := tfPluginClient.ContractsGetter.ListContractsOfProjectName(nodeGroup)
+			if err != nil {
+				trial++
+				log.Debug().Err(err).Str("node group", nodeGroup).Msg("couldn't list contracts")
+				return retry.RetryableError(err)
+			}
+
+			info, err := loadContractsInfo(ctx, tfPluginClient, nodeGroup, contracts.NodeContracts)
+			if err != nil {
+				trial++
+				log.Debug().Err(err).Str("node group", nodeGroup).Msg("couldn't load from grid")
+				return retry.RetryableError(err)
+			}
+
+			nodeGroupsInfo[nodeGroup] = info
+			return nil
+		}); err != nil {
+			failedGroups[nodeGroup] = err.Error()
+		}
+	}
+
+	return parseDeploymentData(nodeGroupsInfo, failedGroups, output)
+}
+
+func loadContractsInfo(ctx context.Context, tfPluginClient deployer.TFPluginClient, nodeGroup string, contracts []graphql.Contract) ([]vmOutput, error) {
+	vmsInfo := []vmOutput{}
+	var multiErr error
+	var lock sync.Mutex
+	var wg sync.WaitGroup
+
+	// Create a channel to act as a semaphore with a capacity of maxGoroutinesToFetchState
+	sem := make(chan struct{}, maxGoroutinesToFetchState)
+
+	for _, contract := range contracts {
+		wg.Add(1)
+
+		// Acquire a slot in the semaphore before starting the goroutine
+		sem <- struct{}{}
+
+		go func(contract graphql.Contract) {
+			defer wg.Done()
+			// Ensure the slot is released as soon as the goroutine completes or errors
+			defer func() { <-sem }()
+
+			deployment, err := workloads.ParseDeploymentData(contract.DeploymentData)
+			if deployment.Type != "vm" {
+				return
+			}
+
+			if err != nil {
+				log.Debug().Err(err).
+					Str("node group", nodeGroup).
+					Msg("couldn't parse deployment data")
+
+				lock.Lock()
+				multiErr = multierror.Append(multiErr, err)
+				lock.Unlock()
+				return
+			}
+
+			log.Debug().
+				Str("vm", deployment.Name).
+				Msg("loading vm info from state")
+
+			vmDeployment, err := tfPluginClient.State.LoadDeploymentFromGrid(ctx, contract.NodeID, deployment.Name)
+			if err != nil {
+				lock.Lock()
+				multiErr = multierror.Append(multiErr, err)
+				lock.Unlock()
+
+				log.Debug().Err(err).
+					Str("vm", deployment.Name).
+					Str("deployment", deployment.Name).
+					Uint32("node ID", contract.NodeID).
+					Msg("couldn't load from state")
+				return
+			}
+
+			vm := vmDeployment.Vms[0]
```

Comment:
Is that because by design the contract has just one vm? even if that was the case, I'd rather to iterate on the vmDeployment.Vms even if it has just 1 item.